File: 508. Most Frequent Subtree Sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var findFrequentTreeSum = function(root) {
    const map = {};
    let maxValueCount = 0;
    (function getFrequent(node) {
        if (node) {
            const value = node.val + getFrequent(node.left) + getFrequent(node.right);
            if (map[value] === undefined) {
                map[value] = 0;
            }
            map[value]++;
            if (map[value] > maxValueCount) {
                maxValueCount = map[value];
            }
            return value;
        }
        return 0;
    })(root);
    return Object.keys(map).filter(i => map[i] === maxValueCount);
};

File: Golang/Tree/508.Most-frequent-subtree-sum.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func findFrequentTreeSum(root *TreeNode) []int {
	hashmap := map[int]int{}
	var helper func(node *TreeNode) int
	helper = func(node *TreeNode) int {
		if node == nil {
			return 0
		}
		value := node.Val + helper(node.Left) + helper(node.Right)
		hashmap[value]++
		return value
	}
	helper(root)
	frequencyHashmap := map[int][]int{}
	max := 0
	for k, v := range hashmap {
		if _, ok := frequencyHashmap[v]; !ok {
			frequencyHashmap[v] = []int{}
		}
		if max < v {
			max = v
		}
		frequencyHashmap[v] = append(frequencyHashmap[v], k)
	}
	return frequencyHashmap[max]
}
